### DISCUSSION TITLE
Orin Nano with U-boot or Grub?

### DISCUSSION THREAD

**Discussion by mickeyproc** [OE4T/meta-tegra#1480] [Q&A]
As far as i understand, there's no support for u-boot for the Orin Nano.
I can however see from the nvidia docs for JP5 that there _is_ support for grub: https://docs.nvidia.com/jetson/archives/r35.4.1/DeveloperGuide/text/SD/Bootloader/UEFI.html#grub-support

I've tried doing a bit of testing and i got the L4TLauncher UEFI bootloader to show "L4TLauncher: Attempting GRUB Boot" but then goes straight to "L4TLauncher: Attempting Direct Boot" and boots on the rootfs.

I've tried adding some entries to the `/etc/grub.d/40_custom`:

```
CMDLINE="mminit_loglevel=4 console=ttyTCU0,115200 firmware_class.path=/etc/firmware fbcon=map:0 net.ifnames=0 nospectre_bhb"

menuentry 'Jetson Linux Recovery' {
        echo "Booting Jetson Linux RECOVERY..."
        linux   /boot/Image root=/dev/disk/by-partlabel/RECOVERY ${CMDLINE} rauc.slot=RECOVERY
        initrd  /boot/initrd
}

menuentry 'Jetson Linux Rootfs.1' {
        echo "Booting Jetson Linux APP..."
        linux   /boot/Image root=/dev/disk/by-partlabel/APP ${CMDLINE} rauc.slot=APP
        initrd  /boot/initrd
}

menuentry 'Jetson Linux Rootfs.2' {
        echo "Booting Jetson Linux APP_b..."
        linux   /boot/Image root=/dev/disk/by-partlabel/APP_b ${CMDLINE} rauc.slot=APP_b
        initrd  /boot/initrd
}
```

I've got the `grub-efi-bootaa64.efi` into the `/opt/nvidia/esp/` folder also.
It doesn't really seem to work though. Nothing really happens other than it's "Attempting GRUB".
Does meta-tegra have support for grub usage and I'm missing something and/or have anyone succeeded with this?

**Comment by dwalkes** [comment]
There have been [discussions about GRUB support](https://github.com/search?q=repo%3AOE4T%2Fmeta-tegra+GRUB&type=wikis) but as far as I know it's not currently implemented (see note [here](https://github.com/OE4T/meta-tegra/wiki/JetPack-5.1.2-L4T-R35.4.1-Notes#other-notes)).

**Reply by mickeyproc** [reply]
Yeah i saw those posts, but that just insinuates that it's doable. I did manage to get GRUB working on the target, though with a bit of UEFI shell managing post-flashing. So it IS doable with some tweaking. I just need to do all the post-flashing steps done in Yocto instead. I had to do the following steps: 

1. Populate the `boot/` dir with kernel image.
2. Add grub-efi-bootaa64.efi file to ESP partition
3. Create GRUB as a boot option in UEFI and move it to the top 
4. Create menu entrie(s) for the bootable partition(s) (rootfs A/B and recovery).

**Comment by redeexpressos** [comment]
Any update on this?

**Reply by mickeyproc** [reply]
Another issue came in with higher priority. However, I am currently looking into using a more barebone custom uefi (edk2) bootloader without l4tlauncher. So when it comes to GRUB support (which is technically the discussion topic), I'm _not_ investigating for now.

However the short time I spend on it, you should be able to just make a recipe or bbappend that populates the ESP partition with a grub efi image instead of the l4tlauncher efi image. And then have a correct grub.conf to specify the boot options and so on. There might be some update usability that is lost with this, that you might want to look at as well as other currently unknown issues (l4tlauncher is removed after all when doing this).

Please feel free to update the discussion thread if you find out more on this.
